styx-common/src/main/java/com/spotify/styx/api/Middlewares.java

```diff
@@ -79,39 +90,94 @@ private Middlewares() {
-            .completedFuture(Response.forStatus(Status.NOT_ACCEPTABLE.withReasonPhrase(
+        return
+            completedFuture(Response.forStatus(Status.NOT_ACCEPTABLE.withReasonPhrase(
                 "blacklisted client version, please upgrade")));
       } else {
-        // noinspection unchecked
-        return (CompletionStage<Response<ByteString>>) innerHandler.invoke(requestContext);
+        return innerHandler.invoke(requestContext);
       }
     };
   }
 
-  public static Middleware<AsyncHandler<? extends Response<?>>,
-      AsyncHandler<? extends Response<ByteString>>> auditLogging() {
+  private static final GoogleIdTokenVerifier GOOGLE_ID_TOKEN_VERIFIER;
+
+  static {
+    final NetHttpTransport transport;
+    try {
+      transport = GoogleNetHttpTransport.newTrustedTransport();
+    } catch (GeneralSecurityException | IOException e) {
+      throw new RuntimeException(e);
+    }
+    GOOGLE_ID_TOKEN_VERIFIER = new GoogleIdTokenVerifier
+        .Builder(transport, Utils.getDefaultJsonFactory())
+        .build();
+  }
+
+  public interface AuthContext {
+    Optional<GoogleIdToken> user();
+  }
+
+  public interface Authenticated<T> extends Function<AuthContext, T> {}
+  public interface Requested<T> extends Function<RequestContext, T> {}
+
+  public static <T> Middleware<Requested<Authenticated<T>>, AsyncHandler<Response<ByteString>>> authed() {
+    return ar -> jsonAsync().apply(requestContext -> {
+      final T payload = ar
+          .apply(requestContext)
+          .apply(auth(requestContext));
+      return completedFuture(Response.forPayload(payload));
+    });
+  }
+
+  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> authValidator() {
+    return h -> rc -> {
+      if (!auth(rc).user().isPresent()) {
+        return completedFuture(Response.forStatus(Status.UNAUTHORIZED));
+      }
+
+      return h.invoke(rc);
+    };
+  }
+
+  public static AuthContext auth(RequestContext requestContext) {
+    return () -> requestContext.request()
+        .header(HttpHeaders.AUTHORIZATION)
+        .filter(s -> s.startsWith(BEARER_PREFIX))
+        .map(s -> s.substring(BEARER_PREFIX.length()))
+        .map(Middlewares::verifyIdToken)
+        .filter(Objects::nonNull);
+  }
+
+  private static GoogleIdToken verifyIdToken(final String s) {
+    try {
+      return GOOGLE_ID_TOKEN_VERIFIER.verify(s);
+    } catch (GeneralSecurityException | IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> auditLogger() {
     return innerHandler -> requestContext -> {
       final Request request = requestContext.request();
+      final AuthContext authContext = auth(requestContext);
       if (!"GET".equals(request.method())) {
-        LOG.info("[AUDIT] {} {} with headers {} parameters {} and payload {}",
+        LOG.info("[AUDIT] {} {} from {} with headers {} parameters {} and payload {}",
```

Comment:
On the wording, I'd rather make `from x` mean REMOTE_ADDRESS or whatever the name for remote ip is and another word for the user. Perhaps `by foo@bar.com`?